### PR TITLE
add mobilenet v3

### PR DESCRIPTION
**Pull request by brettkoonce**
the first commit is the one in #408, was reusing block names and so this requires to be built on top of that branch.

### CODE REVIEW

**Comment by Kshitij09** [comment]
I was also planning to contribute this one :sweat_smile: even further, I had also planned to contribute `mobilenetv2`. Now I'll at least try to come up with examples using these models

**Comment by brettkoonce** [comment]
this requires #408 to be merged first, otherwise there will be a namespace collision

**Comment by brettkoonce** [comment]
also there are some bugs in this implementation currently

**Comment by brettkoonce** [comment]
@texasmichelle sorry for the confusion, this is much better. will bang on the filter resizing logic next.

**Comment by brettkoonce** [comment]
see also https://github.com/keras-team/keras-applications/pull/176/files

**Comment by brettkoonce** [comment]
and https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet_v3.py

**Comment by brettkoonce** [comment]
and https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/mobilenetv3.py

**Comment by brettkoonce** [comment]
@texasmichelle filter resizing is working, let me know what you think!

**Comment by texasmichelle** [comment]
Excellent and thanks for including those links! I'll take a look in the next few days.

File: Models/ImageClassification/MobileNetV3.swift [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+            padding: .same)
+        classiferConv = Conv2D<Float>(
+            filterShape: (1, 1, lastPointChannel, classCount),
+            strides: (1, 1),
+            padding: .same)
+    }
+
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        let initialConv = hardSwish(input.sequenced(through: inputConv, inputConvBatchNorm))
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: updated!

File: Models/ImageClassification/MobileNetV3.swift [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+            invertedResidualBlock5)
+        let backbone2 = backbone1.sequenced(
+            through: invertedResidualBlock6, invertedResidualBlock7,
+            invertedResidualBlock8, invertedResidualBlock9, invertedResidualBlock10,
+            invertedResidualBlock11)
+        let outputConvResult = hardSwish(outputConvBatchNorm(outputConv(backbone2)))
+        let averagePool = avgPool(outputConvResult).reshaped(to: [
+            input.shape[0], 1, 1, lastConvChannel
+        ])
+        let finalConvResult = hardSwish(finalConv(averagePool))
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: added w/ default 0.2 from keras imp

File: Models/ImageClassification/MobileNetV3.swift [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+            padding: .same)
+        classiferConv = Conv2D<Float>(
+            filterShape: (1, 1, lastPointChannel, classCount),
+            strides: (1, 1),
+            padding: .same)
+    }
+
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        let initialConv = hardSwish(input.sequenced(through: inputConv, inputConvBatchNorm))
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: updated!

File: Models/ImageClassification/MobileNetV3.swift [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+            invertedResidualBlock8, invertedResidualBlock9, invertedResidualBlock10)
+        let backbone3 = backbone2.sequenced(
+            through: invertedResidualBlock11,
+            invertedResidualBlock12, invertedResidualBlock13, invertedResidualBlock14,
+            invertedResidualBlock15)
+        let outputConvResult = hardSwish(outputConvBatchNorm(outputConv(backbone3)))
+        let averagePool = avgPool(outputConvResult).reshaped(to: [
+            input.shape[0], 1, 1, self.lastConvChannel
+        ])
+        let finalConvResult = hardSwish(finalConv(averagePool))
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: added w/ default 0.2 from keras imp

File: Models/ImageClassification/MobileNetV3.swift [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+            expansionFactor: 6, seLayer: true, hSwish: true)
+        invertedResidualBlock12 = InvertedResidualBlock(
+            filters: (112, 112), widthMultiplier: widthMultiplier,
+            expansionFactor: 6, seLayer: true, hSwish: true)
+        invertedResidualBlock13 = InvertedResidualBlock(
+            filters: (112, 160), widthMultiplier: widthMultiplier,
+            expansionFactor: 6, kernel: (5, 5), seLayer: true, hSwish: true)
+        invertedResidualBlock14 = InvertedResidualBlock(
+            filters: (160, 160), widthMultiplier: widthMultiplier,
+            expansionFactor: 6, strides: (2, 2), kernel: (5, 5), seLayer: true, hSwish: true)
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: good catch, fixed!

File: Models/ImageClassification/MobileNetV3.swift [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+        batchNormConv2 = BatchNorm(featureCount: filterMult.1)
+    }
+
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        var depthwise = batchNormDConv(dConv(input))
+        if self.useHardSwish {
+            depthwise = hardSwish(depthwise)
+        } else {
+            depthwise = relu(depthwise)
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift:95]: the relu6 is in [hardSwish](https://github.com/tensorflow/swift-apis/blob/27f97e044ac09462e363957c31ab1ae4e1810e44/Sources/TensorFlow/Operators/Math.swift#L1427), is that what you are thinking of?
everything else seems to point to using relu: upstream keras [here](https://github.com/keras-team/keras-applications/blob/bc89834ed36935ab4a4994446e34ff81c0d8e1b7/keras_applications/mobilenet_v3.py#L514) and [here](https://github.com/tensorflow/models/blob/f2dcb8d4bf76192f3cb0730c29cf25a9b1f92a31/research/slim/nets/mobilenet/mobilenet_v3.py#L195), table 1 + 2 of [paper](https://arxiv.org/pdf/1905.02244.pdf), tensorflow slim [model](https://github.com/tensorflow/models/blob/f2dcb8d4bf76192f3cb0730c29cf25a9b1f92a31/research/slim/nets/mobilenet/mobilenet_v3.py#L61).

File: Models/ImageClassification/MobileNetV3.swift [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+    return (
+        makeDivisible(filter: filters.0, widthMultiplier: widthMultiplier),
+        makeDivisible(filter: filters.1, widthMultiplier: widthMultiplier)
+    )
+}
+
+public struct InitialInvertedResidualBlock: Layer {
+    @noDerivative public var addResLayer: Bool
+    @noDerivative public var useSELayer: Bool = false
+    @noDerivative public var useHardSwish: Bool = false
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: see updated!

File: Models/ImageClassification/MobileNetV3.swift [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+    public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+        var depthwise = batchNormDConv(dConv(input))
+        if self.useHardSwish {
+            depthwise = hardSwish(depthwise)
+        } else {
+            depthwise = relu(depthwise)
+        }
+        var squeezeExcite: Tensor<Float>
+        if self.useSELayer {
+            let seAvgPoolReshaped = seAveragePool(depthwise).reshaped(to: [
```
[the remark this answers is not in the document]
Comment (brettkoonce) [reply Models/ImageClassification/MobileNetV3.swift]: see updated!

**Comment by brettkoonce** [comment]
@texasmichelle running regression tests, see note above, should be ready to go!

**Pull request merged**